File: services/eval/src/metrics/ndcg.ts
export function ndcgAtK(retrievedIds: string[], relevantIds: string[], k: number): number {
  if (k <= 0) throw new Error(`k must be positive, got ${k}`);
  if (relevantIds.length === 0) return 0;

  const relevantSet = new Set(relevantIds);
  const topK = retrievedIds.slice(0, k);

  const dcg = topK.reduce((acc, id, i) => {
    const gain = relevantSet.has(id) ? 1 : 0;
    return acc + gain / Math.log2(i + 2);
  }, 0);

  const idealHits = Math.min(relevantSet.size, k);
  let idcg = 0;
  for (let i = 0; i < idealHits; i++) {
    idcg += 1 / Math.log2(i + 2);
  }

  if (idcg === 0) return 0;
  return dcg / idcg;
}

File: services/eval/src/rag_eval/metrics/ndcg.py
import math
from collections.abc import Sequence


def ndcg_at_k(retrieved_ids: Sequence[str], relevant_ids: Sequence[str], k: int) -> float:
    """Normalized discounted cumulative gain at k with binary relevance."""
    if k <= 0:
        raise ValueError(f"k must be positive, got {k}")
    if len(relevant_ids) == 0:
        return 0.0

    relevant_set = set(relevant_ids)
    top_k = retrieved_ids[:k]

    dcg = sum(
        (1 if chunk_id in relevant_set else 0) / math.log2(position + 2)
        for position, chunk_id in enumerate(top_k)
    )

    ideal_hits = min(len(relevant_set), k)
    idcg = sum(1 / math.log2(position + 2) for position in range(ideal_hits))

    if idcg == 0:
        return 0.0
    return dcg / idcg
